urllib3/connection.py

```diff
@@ -299,7 +302,15 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+            try:
+                match_hostname(cert, self.assert_hostname or hostname)
+            except CertificateError as e:
+                log.error(
+                    'Certificate did not match expected hostname: %s. '
+                    'Certificate: %s', asserted_hostname, cert
+                )
+                e.cert = cert
+                raise e
```

Comment:
I'm a *little* nervous about the fact that this exception will have a misleading traceback. We might need to do something a bit nasty here to preserve the original traceback, rather than having it be emitted from here. I think [this code](https://stackoverflow.com/questions/8760267/re-raise-python-exception-and-preserve-stack-trace) will do the job, but I admit to not having tried.